### PR TITLE
Add DMI Bridge

### PR DESCRIPTION
**Pull request by soohyuk-cho**
This adds a DMI bridge and associated collateral. Currently, it doesn't batch DMI requests. Same support as the TSI interface. Also includes a variety of small fixes in other places.

Tested in the past to boot Linux (slowly since there are no batched DMI requests).

If the DMI/TSI bridge are connected, this leads to undefined behavior (same behavior as Chipyard).

### Contributor Checklist
- [ ] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous prints/debugging code?
- [ ] Did you state the UI / API impact?
- [ ] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#include "firesim_dtm.h"
+#include <inttypes.h>
+#include <stdio.h>
+
+#define fprintf(stdout, fmt, ...) (0)
+
+firesim_dtm_t::firesim_dtm_t(int argc, char **argv, bool can_have_loadmem)
+    : testchip_dtm_t(argc, argv, can_have_loadmem), is_busy(false),
+      is_loaded_in_host(false), is_loaded_in_target(false) {
+  idle_counts = 10;
```
Comment (joonho3020) [inline sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc:11]: @abejgonzalez Can you remind me how this number was decided?

File: sim/src/main/makefrag/firesim/metasim.mk [hunk 2]
```diff
@@ -46,31 +46,31 @@ sim_binary_basename := $(basename $(notdir $(SIM_BINARY)))
 run-verilator: $(verilator)
 	cd $(dir $<) && \
 	$(verilator) +permissive $(verilator_args) $(COMMON_SIM_ARGS) $(MIDAS_LEVEL_SIM_ARGS) $(EXTRA_SIM_ARGS) +permissive-off $(abspath $(SIM_BINARY)) </dev/null \
-	$(disasm) $(sim_binary_basename).out
+	$(disasm) $(firesim_base_dir)/$(sim_binary_basename).out
```
Comment (joonho3020) [inline sim/src/main/makefrag/firesim/metasim.mk:49]: why is this added?

File: sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc (hunk 1 above)
Comment (abejgonzalez) [reply sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc:11]: The `firesim_dtm_t` class is mainly a copy of the corresponding `*tsi_t` version. So that member is a copy of the same member of that class. It is a legacy member that was put in the original FireSim release. From what I can see, it's just used for the blockdevice bridge to save on MMIO calls if not needed.

Comment (joonho3020) [reply sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc:11]: Oh I meant why idle_count is set to 10, not some other number?

Comment (abejgonzalez) [reply sim/firesim-lib/src/main/cc/fesvr/firesim_dtm.cc:11]: That I don't know. I think its arbitrary

**Review by joonho3020 — APPROVED** [review]
SFTM
